src/Files.App/Shell/ThreadWithMessageQueue.cs

```diff
@@ -1,59 +1,63 @@
 ﻿using System;
-using System.Collections.Concurrent;
 using System.Runtime.Versioning;
 using System.Threading;
+using System.Threading.Channels;
 using System.Threading.Tasks;
 
 namespace Files.App.Shell
 {
 	[SupportedOSPlatform("Windows")]
 	public class ThreadWithMessageQueue : Disposable
 	{
-		private readonly BlockingCollection<Internal> messageQueue;
+		private readonly Channel<Internal> channel;
 
 		private readonly Thread thread;
 
 		protected override void Dispose(bool disposing)
 		{
 			if (disposing)
 			{
-				messageQueue.CompleteAdding();
+				channel.Writer.TryComplete();
 				thread.Join();
-				messageQueue.Dispose();
 			}
 		}
 
 		public async Task<V> PostMethod<V>(Func<object> payload)
 		{
 			var message = new Internal(payload);
-			messageQueue.TryAdd(message);
+
+			if (await channel.Writer.WaitToWriteAsync())
+				channel.Writer.TryWrite(message);
 
 			return (V)await message.tcs.Task;
 		}
 
-		public Task PostMethod(Action payload)
+		public async Task<object> PostMethod(Action payload)
 		{
 			var message = new Internal(payload);
-			messageQueue.TryAdd(message);
 
-			return message.tcs.Task;
+			if (await channel.Writer.WaitToWriteAsync())
+				channel.Writer.TryWrite(message);
+
+			return await message.tcs.Task;
 		}
 
 		public ThreadWithMessageQueue()
 		{
-			messageQueue = new BlockingCollection<Internal>(new ConcurrentQueue<Internal>());
+			channel = Channel.CreateUnbounded<Internal>(new UnboundedChannelOptions { SingleReader = true });
 
-			thread = new Thread(new ThreadStart(() =>
+			thread = new Thread(new ThreadStart(async () =>
 			{
-				foreach (var message in messageQueue.GetConsumingEnumerable())
+				while (await channel.Reader.WaitToReadAsync())
 				{
-					var res = message.payload();
-					message.tcs.SetResult(res);
+					while (channel.Reader.TryRead(out var message))
+					{
+						var res = message.payload();
+						message.tcs.SetResult(res);
+					}
 				}
 			}));
 
-			thread.SetApartmentState(ApartmentState.STA);
```

Comment:
Note that this is important. Some menu items won't appear this way. If you find that this is not true, then we could just stop using ThreadWithMessageQueue alltogether.

Reply:
If this code is left, the exception below is thrown when right-clicking two or three times.
![image](https://user-images.githubusercontent.com/66369541/234733063-fa2e6e37-cbee-493a-8683-41e36f854708.png)

Should we stop usign ThreadWithMessageQueue?